### DISCUSSION TITLE
Allow setting/Detect track number during tagging

### DISCUSSION THREAD

**Discussion by martimpassos** [beetbox/beets#4318] [Ideas]
When an album is incomplete or has shuffled tracks, I'm forced to `abort` the tagging so I don't mess up my tags (something that would be really difficult to detect later, specially when dealing with mixes of the same song):

<img width="720" alt="Screen Shot 2022-03-15 at 18 24 31" src="https://user-images.githubusercontent.com/49573683/158476383-8672b666-07b7-4bfc-9668-132bf1dc5873.png">

In the example above you can clearly see that the correct track is `#2`, but it would be overwritten it as the "Space Remix".

I understand that beets already tries to match track names based on similarity, so I'm curious about why it is failing here. Is it because of `length`? An option `Reorder tracks` would allow working around this though, maybe printing each existing track and asking which index it corresponds to in the selected suggestion? I think the `edit` plugin works along these lines, but unfortunately I haven't been able to use it (I suspect it has to do with my Google Drive music collection being mounted via `rclone mount`).

**Comment by martimpassos** [comment]
Another example of a match so close, yet so far:

<img width="786" alt="Screen Shot 2022-03-16 at 00 44 22" src="https://user-images.githubusercontent.com/49573683/158513940-72c4da3e-7ccd-478c-8931-b1230992290c.png">

**Comment by JOJ0** [comment]
Hi, I am often facing similar issues. Usually it happens with badly or entirely untagged tracks. Here is an example:

```
Correcting tags from:
    (unknown album)
To:
    Kaotic Chemistry - LSD EP
URL:
    https://musicbrainz.org/release/f7f57846-a6c5-4276-833b-a470bf38b3f5
(Similarity: 26.3%) (tracks, album, artist) (12" Vinyl, 1992, GB, Moving Shadow, SHADOW 20)
 * SHADOW020-B1-Kaotic Chemistry-Illegal Subs - F (#0)  -> Space Cakes (#1) (title)
 * SHADOW020-B2-Kaotic Chemistry-Drum Trip II - Am (#0) -> L.S.D. (#2) (id, title)
 * SHADOW020-A1-Kaotic Chemistry-Space Cakes - Am (#0)  -> Illegal Subs (#3) (id, title)
 * SHADOW020-A2-Kaotic Chemistry-L.S.D. - Bbm (#0)      -> Drum Trip II (#4) (title)
Apply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, plaY, eDit, edit Candidates?
```

These mp3's are entirely untagged, so original track number shows `#0'.

What I am always wondering in this case is: Would in such a case, let's call it "untagged, beets has no idea of track ordering from reading tags alone" a possible solution be to at least try to use the order the files are named?

In this case (and in most cases of my badly tagged files) there is a proper numbering lying in the files, either Vinyl-typical or digital file typical. E.g:

```
-rwx------  1 jojo  staff     10786 Jan  3  2007 SHADOW020 A.jpg
-rwx------  1 jojo  staff     21120 Jan  3  2007 SHADOW020 B.jpg
-rwx------@ 1 jojo  staff  11434975 Mar 22 11:36 SHADOW020-A1-Kaotic Chemistry-Space Cakes.mp3
-rwx------@ 1 jojo  staff   9482907 Mar 22 11:36 SHADOW020-A2-Kaotic Chemistry-L.S.D..mp3
-rwx------@ 1 jojo  staff  12003181 Mar 22 11:36 SHADOW020-B1-Kaotic Chemistry-Illegal Subs.mp3
-rwx------@ 1 jojo  staff  11049688 Mar 22 11:36 SHADOW020-B2-Kaotic Chemistry-Drum Trip II.mp3
-rwx------  1 jojo  staff        64 Jan  3  2007 SHADOW020.txt
```

So my specific question to @sampsyo would be: Do you think it's difficult/doable to code a "fallback sorting" mechanism "as on-disk", in case no track numbering can be found in the tags at all?

**Reply by JOJ0** [reply]
Following up here. In the case that at least a title tag is present in the original files, the ordering is fine:


```
    https://musicbrainz.org/release/b380e498-1ed8-4cc4-9cef-52c41b695e1a
(Similarity: 42.9%) (album, artist, tracks) (12" Vinyl, 1992, GB, Moving Shadow, SHADOW 21)
 * x-Tone Def - Projection - Abm (#0)    -> Projection (#1) (title)
 * xx-Tone Def - Aftertouch - Dm (#0)    -> Aftertouch (#2) (title)
 * y-Tone Def - Power Metric - Fm (#0)   -> Powermetric (#3) (title)
 * yy-Tone Def - Psychic T.D. - Abm (#0) -> Psychic T.D. (#4) (title)
Apply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, plaY, eDit, edit Candidates?
```

Here, the original files are badly tagged as well but _do have_ one tag filled: title.

Am I guessing right that the ordering in this case is not messed up because it alphanumerically is sorted by the data in title? Or is it a coincidence?

**Comment by JOJ0** [comment]
> I understand that beets already tries to match track names based on similarity, so I'm curious about why it is failing here. Is it because of `length`? An option `Reorder tracks` would allow working around this though, maybe printing each existing track and asking which index it corresponds to in the selected suggestion? I think the `edit` plugin works along these lines, but unfortunately I haven't been able to use it (I suspect it has to do with my Google Drive music collection being mounted via `rclone mount`).

And yes, certainly this is also what I am thinking each time the track ordering is messed up. How could a "reording" mechanism work during the import or is there one already I am not aware of?

What I tried:
- Manually quickly retagging the original files. Just giving them the track numbering and nothing else. That doesn't work because there is no way in the importer to tell "re-read tags from disk". And it involves a third party tagging tool.
- Using the "Edit Candidates" function. There unfortunately is no way to match "filenames" to "fetched tags".

**Reply by martimpassos** [reply]
> What I tried:
> 
> Manually quickly retagging the original files. Just giving them the track numbering and nothing else. That doesn't work because there is no way in the importer to tell "re-read tags from disk". And it involves a third party tagging tool.

I suggested something similar [here](https://discourse.beets.io/t/editing-moving-files-externally-during-import/2058). Have you tried the [fromfilename](https://beets.readthedocs.io/en/stable/plugins/fromfilename.html) plugin? It could help for files containing the track number, but I suppose they'd need to include artist and album as well.

> And yes, certainly this is also what I am thinking each time the track ordering is messed up. How could a "reording" mechanism work during the import or is there one already I am not aware of?

Related discussion [here](https://github.com/beetbox/beets/discussions/4320#discussioncomment-2374014), appreciate your +1 there :)

**Reply by JOJ0** [reply]
> Have you tried the [fromfilename](https://beets.readthedocs.io/en/stable/plugins/fromfilename.html) plugin? It could help for files containing the track number, but I suppose they'd need to include artist and album as well.

How are you handling/configuring the fromfilename plugin? Do you disable all other metadata sources plugins in the config prior to using it. I can't find a way to make use of it without having to fiddle around with `beet config -e` first :-/

I tried to use source weight like this but I am not sure if it ever worked:
```
fromfilename:
    source_weight: 2.0
    ...
musicbrainz:
    source_weight: 1.0
    ...
discogs:
    source_weight: 0.7
    ...
```

From what I "heard",  fromfilename only works if _no tags are present at all_ so I figured my config above should do what I want: "Fall back to it when there is no tags at all" but I never find anything in the verbose log mentioning that it is working with fromfilename. I was too lazy to read code/investigate yet. Let's share how to handle this plugin here :-) Thanks!

**Reply by JOJ0** [reply]
I gave fromfilename another try since I just had an entirely untagged file (had ripped it from Vinyl myself):
```
$ beet -vvv import --incremental ~/Music/Techno/031-spin_14-04/Josh\ Wink\ -\ Are\ You\ There\ Remixes\ \[OVM-242\]\ \(2014\)/
user configuration: /Users/jojo/.config/beets/config.yaml
data directory: /Users/jojo/.config/beets
plugin paths: /Users/jojo/git/mried_beetsplug/beetsplug; /Users/jojo/git/whatlastgenre/plugin/beets/beetsplug
artresizer: method is (2, (7, 0, 7), False)
thumbnails: using IM to write metadata
thumbnails: using Python Pathlib to compute URIs
Sending event: pluginload
library database: /Users/jojo/.beet/jtbeets_gin.db
library directory: /Users/jojo/Music/Techno
Sending event: library_opened
Sending event: import_begin
ignoring b'.DS_Store' due to ignore rule b'.*'
Sending event: import_task_created
importadded: In place import detected, recording mtimes from source paths
importadded: Recorded mtime 1648033019.954088 for item '/Users/jojo/Music/Techno/031-spin_14-04/Josh Wink - Are You There Remixes [OVM-242] (2014)/01. Josh Wink - Are You There (Ben Klock Remix).mp3' imported from '/Users/jojo/Music/Techno/031-spin_14-04/Josh Wink - Are You There Remixes [OVM-242] (2014)/01. Josh Wink - Are You There (Ben Klock Remix).mp3'
Sending event: import_task_start
chroma: chroma: fingerprinted b'/Users/jojo/Music/Techno/031-spin_14-04/Josh Wink - Are You There Remixes [OVM-242] (2014)/01. Josh Wink - Are You There (Ben Klock Remix).mp3'
chroma: no match found
Looking up: /Users/jojo/Music/Techno/031-spin_14-04/Josh Wink - Are You There Remixes [OVM-242] (2014)
Tagging 01. Josh Wink  -
No album ID found.
Search terms: 01. Josh Wink  -
Additional search terms: {'label': '', 'catalognum': ''}
Album might be VA: False
discogs: Searching for master release 44820
discogs: Searching for master release 51719
discogs: Searching for master release 86639
discogs: Searching for master release 86639
Sending event: albuminfo_received
```

**Reply by JOJ0** [reply]
from this line (coming from `autotag/match.py):
```
Tagging 01. Josh Wink  -
```
I suppose that beets is taking `01. Josh Wink`as artist and using it in a search on whatever metadata source plugins are in use.

No sign of fromfilename plugin being passed this info though. 

You have any idea what I'm doing wrong?

**Reply by JOJ0** [reply]
I finally started investigating and it turns out the fromfilename plugin is just very silent. It does not log a single thing. IMHO it would clarify things if at least in verbose imports some essential things would be stated. I started to play around with this and looking forward to finally understanding how this plug is actually working. From what I just got now is that I was assuming wrong (as I wrote above): It does not only get into service when no metadata at all can be found but specifically checks `track #` and `title` fields for potential validity and if it decides they are bad, it replaces them with data from actual file names. Actually pretty cool! Will continue working on improving logging but now have better things to do, since the sun is shining in the country I'm living in ;-))

**Reply by JOJ0** [reply]
State of investigation so far:

One file in folder named:
```
01. Josh Wink - Are You There (Ben Klock Remix).mp3
```
Importing with some new log statements:
```
 $ beet -vv import --incremental ~/Music/Techno/031-spin_14-04/Josh\ Wink\ -\ Are\ You\ There\ Remixes\ \[OVM-242\]\ \(2014\)/
...
fromfilename: Artist replaced with: 01. Josh Wink
fromfilename: Bad title replaced with:  Are You There (Ben Klock Remix)
fromfilename: Track replaced with: 1
...
```

Not too bad of a choice for search terms but the odd thing is: I'm getting bollocks results in the ddiscogs and mb search.
Tried to use exactely the same information (not entirely correct artist name!!), using Ènter Search and voila: It perfectly finds the release I'm looking for. Weird, need to investigate more next time.
